### DISCUSSION TITLE
Help with user filter on view collection, or perhaps to do it differently

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#2370] [Q&A] [answered]
I've created an OR condition where admin users (>= rank 50) can view the records for admin listing.  However then all admin users will get other people's items in their normal user listing.  I'm wondering if `sites_users` could be used in the front end filter to then restrict it down.

2) Alternatively I could make two separate views.  One for admin-only that shows all items, and then another for user displays (which could also be an admin) that only shows the item if you have a record in the many-to-many.

@ganigeorgiev Do you think that the second approach is less problematic?  It would be nice to just have one view, but it's complicated because I'm opening it up to admins with a special condition, and then trying to close it back down again.

To clarify, my "admin" isn't a pocketbase admin.  I could do that, but I really don't want to trust people with schema privileges.

Another idea would be that Pocketbase could have admin users with only limited access to all list and view records, but not anything extra for delete, update and create.  That is probably a huge can of worms though.  I'm leaning towards my 2) alternative.

View SQL

```sql
SELECT
  s.id,
  pInt.name as nameInt,
  pExt.name as nameExt,
  s.type,
  s.status,
  s.created,
  s.updated
FROM sites s
/* internal point */
LEFT JOIN sites_points spInt ON spInt.site = s.id
LEFT JOIN points pInt ON spInt.point = pInt.id
/* external point */
LEFT JOIN sites_points spExt ON spExt.site = s.id
LEFT JOIN points pExt ON spExt.point = pExt.id

WHERE
  s.status = 'active' AND
  spInt.type = 'int' AND
  spExt.type = 'ext'
GROUP BY s.id
```

View filter
```
(
   (@collection.sites_users.site ?= id && 
   @collection.sites_users.user ?= @request.auth.id)
|| type = 'briefcase'
|| @request.auth.group.rank >= 50
)
&& 'active' = status
```

**Comment by ganigeorgiev** [comment]
> I'm wondering if sites_users could be used in the front end filter to then restrict it down.

The `@collection.*` filters can be applied only as part of the API rules (aka. only by the PocketBase admins). You can apply a client-side filter on `sites_users` only if it as a direct collection filter or through a relation field originated from the current list/search collection.

---

> Do you think that the second approach is less problematic?

Unfortunately I don't have the full picture to understand the business requirements or the use cases and I'm not sure what approach to recommend.

> I've created an OR condition where admin users (>= rank 50) can view the records for admin listing. However then all admin users will get other people's items in their normal user listing.

What is the end goal? If you want to restrict your admin users to have access to only their items wouldn't AND (`&&`) grouping the `>=50` condition with the `@collection.*` filters do the job?

**Reply by dschissler** [reply]
> What is the end goal? If you want to restrict your admin users to have access to only their items wouldn't AND (&&) grouping the >=50 condition with the @collection.* filters do the job?

Simpler than that.  If I only want admin (not PB admin) users to be able to view their items then I would just remove that condition.  That would require them to have the many-to-many record in `sites_users` the same as everyone.

I'm heavily leaning towards making separate admin screen collection views.

**Reply by ganigeorgiev** [reply, marked as answer]
I'm still not entirely sure that I understand what you are trying todo, but if you want to have the option to "toggle" the behavior admin users to have access to all items client-side, aka. to maintain a single view but have 2 different list results, you can also use a helper query parameter and check it inside the rules via `@request.query.*`. Something like:

```js
status = 'active' &&
(
   type = 'briefcase'
   ||
   (
      @collection.sites_users.site ?= id && 
      @collection.sites_users.user ?= @request.auth.id
   )
   ||
   (
      @request.query.adminsOnly = 1 &&
      @request.auth.group.rank >= 50
   )
)

```

**Reply by dschissler** [reply]
Thanks!  The `request.query` idea was the missing piece that I needed.

**Reply by dschissler** [reply]
@ganigeorgiev I've been wanting to ask if you came up with this whole SQL with API Filter pattern yourself, or is it borrowed from some other projects that I should know about?

**Reply by ganigeorgiev** [reply]
It was heavily inspired by the [Firebase security rules](https://firebase.google.com/docs/firestore/security/get-started), but I wanted also to be able to combine them with server-side filters (_eg. return only "active" items_) and at the same time to have similar client-side filter syntax.

It definitely requires some time to get used to, but I don't think it is anything too unusual or unique. The syntax basically follows `OPERAND OPERATOR OPERAND` pattern.

**Reply by dschissler** [reply]
I think that the API Rules could use one or two pages of examples.  Perhaps a slimmed down version of this thread could be one of them.  This example does about four things;  It shows the very common pattern of filtering by user using auth.id, expanding a relation directly off of the user auth record using dot notation, and also passing query parameters, and also joining using a typical many-to-many membership table.  It also shows a good formatting style.
